### PR TITLE
sds: certificate hot-reload for xDS gRPC connection

### LINKED ISSUE #9359: Provide support for certificate rotation for xDS connection in Envoy container images
**Problem description**:

The official docker container is used by e.g. Contour ingress controller.

While Contour is able to replace certificates for user plane by using SDS, it is not currently possible to rotate control plane certificates (xDS gRPC interface) without traffic interruption.

**Proposal for a new feature**:

Include a new binary in official Envoy container images at https://hub.docker.com/r/envoyproxy/.  This new binary can be used by deployments to

1) acting as pid 1 in the container and forking & execing `envoy`
2) triggering Envoy's hot restart procedure when certificate or key files are updated

inotify can be used to watch the file updates, which works with e.g. Kubernetes secret volume mounts.

**Implementation alternatives**:

With your guidance, I'd be interested in implementing the above feature and submitting PR.

The default entrypoint can still remain like it is currently.  The new binary could be "opt-in" for deployments that require xDS with TLS and hot-reload.

One implementation alternative would be to extend [hot-restarter.py](https://github.com/envoyproxy/envoy/tree/master/restarter) with inotify, but alternatively new version could be implemented with C++ in order to avoid bringing Python as dependency into all Envoy images.

### PR DESCRIPTION
**Pull request by tsaarni**
Create watches for secrets pointed by path based SDS resources and trigger
hot-reload when the files change.  This allows rotation of TLS certificate
and key for the xDS gRPC connection without restart.

Signed-off-by: Tero Saarni <tero.saarni@est.tech>

**Description:**
This PR adds support for certificate rotation for xDS gRPC interface. The implementation approach has been discussed in https://github.com/envoyproxy/envoy/issues/9359.  

The PR consists of following parts:

* Add capability to SDS to make filesystem subscriptions for the certificate and key files referred to from SDS resources.
* Adds capability to inotify `WatcherImpl` to make subscriptions for directories instead of only files. The change is backwards compatible.  

The latter is needed to support commonly used pattern of updating files with atomic directory swap.  Watching individual files is not possible, since in this pattern the files are symbolic links which point to a directory.  The target of the symbolic links is a directory, which is atomically renamed (replaced) at update, not the files.  Kubernetes uses this approach for secret volume mount. The details are covered in: https://github.com/envoyproxy/envoy/issues/9359#issuecomment-579314094 and by the test case introduced in this PR.

The feature introduced by this PR is needed by Contour https://github.com/projectcontour/contour/issues/2143 but it can be beneficial to any user of Envoy. 

**Risk Level:** Medium
**Testing:** unittest
**Docs Changes:** To be defined: add information about the feature somewhere in the documentation
**Release Notes:**

Fixes #9359

### CODE REVIEW

**Comment by tsaarni** [comment]
I have added the new behavior to only inotify file watcher for Linux, but should this be added to kqueue watcher too -  I guess the test case I added may fail on MacOS?

**Comment by tsaarni** [comment]
After adding kqueue watcher implementation for macos, a different test case failed.  I cannot find direct link between my change and [the new failure](https://github.com/envoyproxy/envoy/pull/10163/checks?check_run_id=478974287).  I see it occasionally failing on my machine too, ranging from 1 failure out of 1000 runs, to 3 out of 1000:

```
bazel test --runs_per_test 1000 //test/extensions/common/redis:cluster_refresh_manager_test`
...
//test/extensions/common/redis:cluster_refresh_manager_test              FAILED in 3 out of 1000 in 2.6s
```

**Comment by tsaarni** [comment]
I'd appreciate review comments on this PR, thank you!

File: source/common/secret/sds_api.h [hunk 1]
```diff
@@ -65,6 +66,16 @@ class SdsApi : public Config::SubscriptionCallbacks {
   }
   static std::string loadTypeUrl(envoy::config::core::v3::ApiVersion resource_api_version);
 
+  virtual std::vector<std::string> getDataSourceFilenames() PURE;
+
+  uint64_t getHashForFiles() {
```
[the remark this answers is not in the document]
Comment (tsaarni) [reply source/common/secret/sds_api.h]: done!

File: source/common/secret/sds_api.h [hunk 2]
```diff
@@ -76,11 +87,15 @@ class SdsApi : public Config::SubscriptionCallbacks {
   const std::string sds_config_name_;
 
   uint64_t secret_hash_;
+  uint64_t files_hash_;
   Cleanup clean_up_;
   ProtobufMessage::ValidationVisitor& validation_visitor_;
   Config::SubscriptionFactory& subscription_factory_;
   TimeSource& time_source_;
   SecretData secret_data_;
+  std::unique_ptr<Filesystem::Watcher> watcher_;
```
[the remark this answers is not in the document]
Comment (tsaarni) [reply source/common/secret/sds_api.h]: done!

File: source/common/secret/sds_api.h [hunk 3]
```diff
@@ -204,6 +238,17 @@ class CertificateValidationContextSdsApi : public SdsApi,
     validation_callback_manager_.runCallbacks(secret.validation_context());
   }
 
+  std::vector<std::string> getDataSourceFilenames() override {
```
[the remark this answers is not in the document]
Comment (tsaarni) [reply source/common/secret/sds_api.h]: done!

**Comment by lizan** [comment]
Can you merge master? thanks

**Comment by tsaarni** [comment]
I've rebased the PR.
There was some flakiness in test environment this time...

**Comment by lizan** [comment]
Sorry for the delay, can you merge master again? not rebase but just merge.

**Comment by tsaarni** [comment]
I merged but got some new errors with the CI https://dev.azure.com/cncf/envoy/_build/results?buildId=36546&view=logs&j=3bf5ec37-bb6f-5fb6-2cf1-5134dc6106d6&t=41d571fe-3811-578d-07de-68468f467dae&l=1567

1. variable `generic_secret` should be `generic_secret_`:  this code existed there before
2. `sys/event.h` file missing: this include is in macos specific code and should not exist on linux

**Comment by lizan** [comment]
@tsaarni I merged this ahead and will fix clang-tidy later. Thanks.

**Comment by mattklein123** [comment]
@lizan @tsaarni this change should have a release note and documentation updates about the new behavior. @tsaarni can you please add some in a follow up? Thank you!

**Comment by tsaarni** [comment]
I propose to:

1. Add new example in [Security / Secret discovery service (SDS)](https://www.envoyproxy.io/docs/envoy/latest/configuration/security/secret) to explain how to use SDS with filesystem based config resources to achieve hot-reload for xDS gRPC.
2. Mention that filesystem subscriptions are also set up for referred certificate and key files in documentation of v2 [core.ConfigSource](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/core/config_source.proto#envoy-api-msg-core-configsource) and [config.core.v3.ConfigSource](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/config_source.proto#config-core-v3-configsource)

Though I'm bit unsure if the 2) is the best place in the API documentation for this information - at least those are the places where it already mentions about filesystem subscriptions in general.

I'm preparing documentation and will be submitting a PR soon for commenting.

**Comment by mattklein123** [comment]
Thank you @tsaarni we can discuss the details in the PR!

**Pull request merged**